Add Torq connector - closed, reopened as #149405

natexcvi opened the pull request:
## Summary

Add a new action type for Torq which triggers Torq workflows.

This was closed and reopened as https://github.com/elastic/kibana/pull/149405 .

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

natexcvi committed:
Initial implementation

natexcvi committed:
Further impl and tests

natexcvi committed:
Add default body

natexcvi committed:
Merge branch 'elastic:main' into add_torq_connector

natexcvi committed:
Change texts + fixes

natexcvi committed:
Merge remote-tracking branch 'origin/add_torq_connector' into add_torq_connector

natexcvi committed:
Merge remote-tracking branch 'origin/main' into add_torq_connector

orihoogi committed:
Type fixes, translations and tests

orihoogi committed:
More fixes

natexcvi committed:
Add default payload and docs

natexcvi committed:
Fix action result debug log

natexcvi committed:
Merge branch 'elastic:main' into add_torq_connector

natexcvi committed:
Revert dependency change

natexcvi committed:
Merge remote-tracking branch 'origin/add_torq_connector' into add_torq_connector

natexcvi committed:
Fixes & refactoring

natexcvi committed:
Update server tests

natexcvi committed:
Remove debug log

natexcvi committed:
Remove TODO

natexcvi committed:
Remove unused code

natexcvi committed:
Remove TODO

natexcvi committed:
Further CR fixes

natexcvi committed:
Fix test

natexcvi committed:
Remove unneeded dep

natexcvi committed:
Merge pull request #1 from natexcvi/add_torq_connector

Add Torq Connector

natexcvi committed:
Merge branch 'elastic:main' into main

kibanamachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

EricDavisX commented:
Hi, @natexcvi I wanted to reach out, is there a point of contact we can talk to about setting up a real 'test instance' of the Torq Security product for on-going end to end validation? We can certainly finish the core connector work first.  Just getting the ball rolling.  Thank you in advance.

natexcvi committed:
Merge branch 'main' into main

natexcvi commented:
> Hi, @natexcvi I wanted to reach out, is there a point of contact we can talk to about setting up a real 'test instance' of the Torq Security product for on-going end to end validation? We can certainly finish the core connector work first. Just getting the ball rolling. Thank you in advance.

Hi @EricDavisX, of course – I've asked our PM @hagaishapira to contact you so everything can be set up. What further steps should we take at this point?

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

semd commented on x-pack/plugins/triggers_actions_ui/public/application/components/builtin_action_types/torq/torq.tsx:
```diff
@@ -0,0 +1,46 @@
+    validationResult.errors = errors;
+    if (!actionParams.body?.length) {
+      errors.body.push(translations.BODY_REQUIRED);
+    }
+    try {
+      JSON.parse(replaceReferencesWithNumbers(actionParams.body || ""));
+    } catch (e) {
+      errors.body.push(translations.INVALID_JSON);
+    }
```
Maybe it is better to check the JSON format only if it is not empty
```suggestion
    } else {
      try {
        JSON.parse(replaceReferencesWithNumbers(actionParams.body));
      } catch (e) {
        errors.body.push(translations.INVALID_JSON);
      }
    }
```

semd commented on x-pack/plugins/triggers_actions_ui/public/application/components/builtin_action_types/torq/torq.tsx:
```diff
@@ -0,0 +1,46 @@
+import { TorqActionParams, TorqConfig, TorqSecrets } from '../types';
+import * as i18n from './translations';
+
+function replaceReferencesWithNumbers(body: string) {
+  return body.replace(/\{\{[.\w]+\}\}/gm, "42");
+}
```
I don't quite get this, why do we need to replace the mustache variables with numbers?

natexcvi replied:
This way we can easily validate the JSON body. E.g.
```json
{
   "key": {{some_number}}
}
```
Is not a valid JSON because the moustache template is for a numeric value (no quotation marks). In order to allow it when validating, and considering the fact that a number is a valid drop-in replacement for any scalar value template, we replace all templated values with numbers for validation.
LMKWYT

stephmilovic commented:
Hi @natexcvi, I will be reviewing this PR. I just had a quick look and I need 2 things:

- Can you please resolve the conflicts?
- We'll need to add function tests. 

The last connector we added was xMatters in PR https://github.com/elastic/kibana/pull/122357 ; the "function test" I mentioned above are in the `x-pack/test/alerting_api_integration` directory, and we'd want the same kind of tests that were added for xMatters, so hopefully you can follow that PR to figure out what to do.  Don't hesitate to ask questions!

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

semd reviewed (commented):
Hi, thanks for the contribution, I wrote a few comments.

The conflicts are caused by a migration of all connectors that happened last week. Now they live in the `stack_connectors` plugin. 
You should be able to move:
- backend in `x-pack/plugin/actions/server` -> `x-pack/plugins/stack_connectors/server`
- UI in `x-pack/plugins/triggers_actions_ui` -> `x-pack/plugins/stack_connectors/public`

semd commented:
@natexcvi, Have you tried to use the Webhook connector with Torq? If so, what are the limitation you hit?

semd commented on docs/management/cases/add-connectors.asciidoc:
```diff
@@ -35,7 +36,7 @@ image::images/cases-connectors.png[]
 . Enter your required settings. Refer to <<resilient-action-type>>,
 <<jira-action-type>>, <<servicenow-action-type>>, <<servicenow-sir-action-type>>,
-<<swimlane-action-type>>, or <<cases-webhook-action-type>> for connector
+<<swimlane-action-type>>, <<torq-action-type>>, or <<cases-webhook-action-type>> for connector
```
The connector does not have Cases feature, it should not be added in this list

natexcvi committed:
Merge remote-tracking branch 'upstream/main' into pr/natexcvi/139635

natexcvi committed:
Move UI connector

natexcvi committed:
Fix doc

natexcvi committed:
Refactoring

natexcvi committed:
Fixes

natexcvi committed:
Fix docs

natexcvi committed:
Validate JSON only if body not empty

natexcvi committed:
Fix test accordingly

natexcvi committed:
Reduce logging

natexcvi committed:
Action fixes

natexcvi committed:
Add functional tests

natexcvi committed:
Merge remote-tracking branch 'upstream/main' into pr/natexcvi/139635

natexcvi commented:
Hi @stephmilovic, thank you for your review. I have fixed the issues you commented on, and added functional tests as requested.
As for your question about the generic webhook integration - that is the experience we currently offer our customers, and we have come to the conclusion it is inferior (ease of use, security) to a dedicated connector.

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Restore param name

natexcvi committed:
Remove debug log

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Remove only from test

natexcvi committed:
Merge remote-tracking branch 'origin/main' into pr/natexcvi/139635

natexcvi committed:
Merge remote-tracking branch 'upstream/main' into pr/natexcvi/139635

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge remote-tracking branch 'upstream/main' into pr/natexcvi/139635

natexcvi committed:
Remove logger arg

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

semd commented:
@elasticmachine merge upstream

kibanamachine commented:
protected branch 'main' check failed:
  At least 1 approving review is required by reviewers with write access.

semd commented on x-pack/plugins/triggers_actions_ui/public/application/sections/action_connector_form/action_form.tsx:
```diff
@@ -352,7 +352,6 @@ export const ActionForm = ({
               />
             );
           }
-
```
could this change be reverted?

semd commented on x-pack/plugins/triggers_actions_ui/public/plugin.ts:
```diff
@@ -149,12 +149,12 @@ interface PluginsStart {
-      PluginsSetup,
-      PluginsStart
-    >
+  CorePlugin<
+  TriggersAndActionsUIPublicPluginSetup,
+  TriggersAndActionsUIPublicPluginStart,
+  PluginsSetup,
+  PluginsStart
+  >
```
could this be reverted?

semd commented:
@natexcvi There are some files that need to be formatted. It can be fixed using:
```
prettier -w x-pack/plugins/stack_connectors/public/connector_types x-pack/plugins/stack_connectors/server/connector_types x-pack/test/alerting_api_integration/common/fixtures/plugins/actions_simulators/server/plugin.ts x-pack/test/alerting_api_integration/common/fixtures/plugins/actions_simulators/server/torq_simulation.ts x-pack/test/alerting_api_integration/security_and_spaces/group2/tests/actions/connector_types/stack/torq.ts
```

semd commented:
buildkite test this

semd commented:
@natexcvi In order to merge we need to be able to test it, can you please email us a working test instance/credentials to `sergi.massaneda@elastic.co`? 
Thanks!

natexcvi committed:
CR and formatting fixes

natexcvi committed:
Test related fixes

natexcvi committed:
i18 fixes

natexcvi committed:
Fix test

natexcvi committed:
Merge remote-tracking branch 'upstream/main' into pr/natexcvi/139635

semd reviewed (commented):
I locally tested the connector (using invalid authentication). Everything in the configuration seems to work properly, I only have a couple of NITs:

- The authentication error in the connector test is not very descriptive:
![torq_auth_err](https://user-images.githubusercontent.com/17747913/198010089-98002b8b-16b8-4190-97e0-b1c60f26ddf5.png)

- The body validation still looks a bit weird to me, it shows the "Body must be a valid JSON." message if the format is not correct. However, the error message disappears if we do something like `"alerts": {{context.alerts}}`, at the same time, the body editor still warns about one error in the content (screenshot) which is a bit incosistent. 
This is related to the `replaceReferencesWithNumbers` function. Are strict number types really needed by the recipient? I don't see any in the default body template.
![body_err](https://user-images.githubusercontent.com/17747913/198010618-42f5e4e5-c67f-48ad-9391-22dde2924da8.png)

Despite these minor things, I would approve the PR if I was able to confirm the connector execution integrates properly with Torq.

hagaishapira commented:
Hi @semd, 
I've just sent you an invite to a tenant where you'll be able to confirm that the connector is fully functional with Torq 🙂 
This access is provided under our MNDA and partnership agreement with Elastic and is provided for testing purposes only.

In the Torq platform you can create a new Elastic integration [here](https://app.torq.io/integrations/elasticsecurity_monitoring). 
Please let @natexcvi and I know if we can provide you with any further assistance in your testing 🙏 .

natexcvi committed:
More informative error message for 404 status

natexcvi commented:
> I locally tested the connector (using invalid authentication). Everything in the configuration seems to work properly, I only have a couple of NITs:
> 
> * The authentication error in the connector test is not very descriptive:
>   ![torq_auth_err](https://user-images.githubusercontent.com/17747913/198010089-98002b8b-16b8-4190-97e0-b1c60f26ddf5.png)
> * The body validation still looks a bit weird to me, it shows the "Body must be a valid JSON." message if the format is not correct. However, the error message disappears if we do something like `"alerts": {{context.alerts}}`, at the same time, the body editor still warns about one error in the content (screenshot) which is a bit incosistent.
>   This is related to the `replaceReferencesWithNumbers` function. Are strict number types really needed by the recipient? I don't see any in the default body template.
>   ![body_err](https://user-images.githubusercontent.com/17747913/198010618-42f5e4e5-c67f-48ad-9391-22dde2924da8.png)
> 
> Despite these minor things, I would approve the PR if I was able to confirm the connector execution integrates properly with Torq.

Hi @semd, I have made the error message clearer for the case you mentioned. As for the body validation - patterns such as `"key": {{value}}` are needed to properly include templated numeric, boolean and object type values, while not giving up validation of the rest of the JSON syntax. Do you know if the editor can be reasonably configured to align its linting with my validation rules?

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

semd commented on x-pack/plugins/stack_connectors/public/connector_types/stack/torq/torq_params.tsx:
```diff
@@ -0,0 +1,46 @@
+          editAction('body', '', index);
+        }
+      }}
+    />
```
I have found a way to not show the error decorations in the JsonEditor, we can overload the default props this way, it is a little bit hacky, but it will work. we show/hide the errors in the editor in accordance with the errors we get from our own validation.
The rest of the options are the same values as defined in the editor by default.


```suggestion
      euiCodeEditorProps={{
        options: {
          renderValidationDecorations: body && errors?.body?.length ? 'on' : 'off',
          lineNumbers: 'on',
          fontSize: 14,
          minimap: {
            enabled: false,
          },
          scrollBeyondLastLine: false,
          folding: true,
          wordWrap: 'on',
          wrappingIndent: 'indent',
          automaticLayout: true,
        },
      }}
    />
```

natexcvi replied:
Applied your solution, thank you!

semd commented:
@natexcvi Thanks for improving the error. 
I have been doing manual testing using the invitation I received, the configuration and the workflow work perfectly. I have only one question regarding authentication: 

The auth header can be parametrized by the user when the integration is created, along with the secret:
![Captura de pantalla 2022-10-28 a les 18 22 35](https://user-images.githubusercontent.com/17747913/198686385-98974a3a-be9f-447d-9404-f739c0c85972.png)

However, in the connector configuration we only ask for the url and the secret, the header name is hardcoded to `X-Torq-Token` in the connector, so I had to check the connector code to know the authentication header name I had to use. If I set it wrong (as in the screenshot) the connector in Elastic only throws auth errors.

Won't it be problematic? will users know what header name they have to use? I don't see any mention of this in the doc.

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Align linter with custom validations

natexcvi committed:
Merge branch 'main' into main

natexcvi commented:
> @natexcvi Thanks for improving the error message. I have been doing manual testing using the invitation I received, the configuration and the workflow work perfectly. I have only one question regarding authentication:
> 
> The auth header can be parametrized by the user when the integration is created, along with the secret:
> 
> ![Captura de pantalla 2022-10-28 a les 18 22 35](https://user-images.githubusercontent.com/17747913/198686385-98974a3a-be9f-447d-9404-f739c0c85972.png)
> 
> However, in the connector configuration we only ask for the url and the secret, the header name is hardcoded to `X-Torq-Token` in the connector, so I had to check the connector code to know the authentication header name I had to use. If I set it wrong (as in the screenshot) the connector in Elastic only throws auth errors.
> 
> Won't it be problematic? will users know what header name they have to use? I don't see any mention of this in the doc.

@semd Seems like the authentication header auto-population feature was not available to your test account, apologies. Should be working now - please let me know if anything is still off.

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

semd commented:
buildkite test this

semd reviewed (approved):
Tested, the header is auto-populated. Everything works fine.
Connector code LGTM.

@elastic/response-ops your turn.

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Fix test

natexcvi committed:
i18n fixes

natexcvi committed:
Fmt fixes

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge remote-tracking branch 'upstream/main' into pr/natexcvi/139635

natexcvi commented:
@semd Hi there, are any further steps needed on our part at this time?

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

semd commented:
Hi @natexcvi,
We need the review of the @elastic/response-ops team. After their approval we'll be able to merge.
I'll ping them internally.

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
Merge branch 'main' into main

pmuellr commented on x-pack/plugins/stack_connectors/server/connector_types/stack/torq/index.ts:
```diff
@@ -0,0 +1,378 @@
+// secrets definition
+export type ActionTypeSecretsType = TypeOf<typeof SecretsSchema>;
+const secretSchemaProps = {
+  token: schema.nullable(schema.string()),
```
It looks like `token` must always be provided, give the validation below.  I think you can just change this to `schema.string()` (remove the `schema.nullable()` wrapper), and then you won't need the extra validation check.

pmuellr commented on x-pack/plugins/stack_connectors/server/connector_types/stack/torq/index.ts:
```diff
@@ -0,0 +1,378 @@
+// params definition
+export type ActionParamsType = TypeOf<typeof ParamsSchema>;
+const ParamsSchema = schema.object({
+  body: schema.maybe(schema.string()),
```
It appears `body` will have to be a parseable JSON string.  I don't think there's a need for the `schema.maybe()` wrapper here, if it's always expected there will be a body.

pmuellr commented on x-pack/plugins/stack_connectors/server/connector_types/stack/torq/index.ts:
```diff
@@ -0,0 +1,378 @@
+        'X-Torq-Token': token || '',
+        'Content-Type': 'application/json',
+      },
+      data: JSON.parse(data || 'null'),
```
I think we'll want to do this `JSON.parse()` in a separate step, before the request.  The reason is, the data may not be valid JSON, and we'll want a clear error message about that, logged to the console and returned in the `message`. field with an explicit error status.

pmuellr commented on x-pack/plugins/stack_connectors/server/connector_types/stack/torq/index.ts:
```diff
@@ -0,0 +1,378 @@
+  return errorResultUnexpectedError(actionId);
+}
+
+// Action Executor Result w/ internationalisation
```
```suggestion
// Action Executor Result
```

(or remove the comment completely, doesn't seem to add much)

pmuellr commented on x-pack/plugins/triggers_actions_ui/public/plugin.ts:
```diff
@@ -345,6 +345,7 @@ export class Plugin
           ...props,
           actionTypeRegistry: this.actionTypeRegistry,
           connectorServices: this.connectorServices!,
+          // getDefaultActionParams: getDefaultsForActionParams,
```
remove if no longer using this

pmuellr reviewed (changes requested):
Made an initial pass, looking at the server-side code.  I suspect we'll want to change the `torqDefaultBody` value, as it's not going to render array values as expected.  Also suggested to parse the JSON body in the executor separately, in case it ends up not being parseable, to generate a clear message about that.  And a few other comments.

natexcvi committed:
Merge branch 'main' into main

natexcvi committed:
CR fixes

natexcvi committed:
Merge branch 'main' into main

natexcvi commented:
Hi @pmuellr, thanks for your review! I have addressed the issues you listed. The major thing left is the handling of the default body template. Would appreciate your input on that. Also wanted to ask what the next steps would be for the PR.

natexcvi committed:
Merge branch 'main' into main

cnasikas commented:
@natexcvi We did a slight modification in the folder structure of the connectors. We removed the `stack`, and `cases` folders and move all connectors to the same folder (`connector_types`). To resolve the conflicts you need to do the same and move your connector to the same level. Sorry for the inconvenience.

pmuellr commented:
> The major thing left is the handling of the default body template. Would appreciate your input on that. Also wanted to ask what the next steps would be for the PR.

Last first :-).  Another quick check, and I think you'll need to add some validation of the URL, as is done below; users can deny-list certain hosts via the [`xpack.actions.allowedHosts` config](https://www.elastic.co/guide/en/kibana/current/alert-action-settings-kb.html#action-settings), so if someone adds the torq host to the deny-list, the connector will generate an error when it's run.  We add this as part of validation, so that it occurs when editing in the UX, creating / updating via HTTP APIs, and when the connector is executed.

https://github.com/elastic/kibana/blob/30b8a2d0ff01ff2011e1de3d97db34adad4b0816/x-pack/plugins/stack_connectors/server/connector_types/webhook/index.ts#L125-L156 

> The major thing left is the handling of the default body template. Would appreciate your input on that.

I'll need to understand more about the purpose of this connector.  I'm guessing the idea is that the body is passed to the "Elastic Security integration" worklow.  Does that integration already exists?  What input does it accept?  Is this connector expected to work with other Torq workflows?

If you want to send ALL the contextual data for the alert to Torq, and then the workflow deal with it, then the body should probably be `{{.}}`, which will render as a JSON string of all the contextual data.  You can see what this looks like by creating a rule and connector, and using `{{.}}` as the "body" of the connector (mail, slack, log, etc).

natexcvi committed:
Merge remote-tracking branch 'upstream/main' into pr/natexcvi/139635

natexcvi committed:
Post merge fixes

natexcvi committed:
Move default body

natexcvi committed:
Formatting

natexcvi commented:
@pmuellr I believe I already have that validation [here](https://github.com/natexcvi/kibana/blob/cdd3f38b06bc036d0613a54e57dcf53999da68b0/x-pack/plugins/stack_connectors/server/connector_types/torq/index.ts#L120). Is that alright or did you have something else in mind?

As for the body template - we want users to have the freedom to template their own payloads in a way that matches the design of their various Torq Workflows. For the default value, I wish to have all data fields. Following your suggestion, I see that when using `{{.}}` for example (same for top level objects such as `{{context}}`) I get an escaped string containing a JSON representation, rather than the JSON representation itself. This means that if a user builds the following body template:
```json
{
    "alert": {{alert}}
}
```
they will get the following payload at runtime:
```json
{
  "alert": {\"id\":\"9b394f1c-d48c-4fbc-95d2-ce13639e56c8\",\"actionGroup\":\"small\",\"actionGroupName\":\"Small t-shirt\",\"flapping\":false}
}
```
which is not valid JSON. The old behaviour of the moustache value injection would put the actual alert JSON object there rather than an escaped string representation, which is the behaviour I want.
Is there a clean way to achieve that now?

pmuellr commented:
> @pmuellr I believe I already have that validation [here](https://github.com/natexcvi/kibana/blob/cdd3f38b06bc036d0613a54e57dcf53999da68b0/x-pack/plugins/stack_connectors/server/connector_types/torq/index.ts#L120). Is that alright or did you have something else in mind?

Perfect - sorry, somehow I missed that.

> As for the body template - we want users to have the freedom to template their own payloads in a way that matches the design of their various Torq Workflows. 

You might want to look at the [tines connector](https://github.com/elastic/kibana/tree/main/x-pack/plugins/stack_connectors/server/connector_types/tines) to see how they dealt with this, especially since they also had a focus on security alerts.  Perhaps there's some "common shape" we can use for these.

Beyond that, there's some flexibility for connectors to deal with JSON using `renderParameterTemplates()` beyond what the torq connector is currently doing - see the elasticsearch index connector as an example: 

https://github.com/elastic/kibana/blob/96b762926b17bcbd4f2e9f68dbef50863b492ecf/x-pack/plugins/stack_connectors/server/connector_types/es_index/index.ts#L143-L159

natexcvi committed:
Fix template

natexcvi commented:
> @natexcvi checking in to see if there's anything you need help with per my previous comment on `renderParameterTemplates()`

Hi, followed your suggestion and it seems to be working as I want using the `escape` arg. Thanks!

natexcvi committed:
Fmt fix

natexcvi committed:
Merge remote-tracking branch 'upstream/main'

natexcvi committed:
Merge branch 'main' into main

shahzad31 commented:
> Apologies for the mess, a result of a botched rebase. Should be good now

Might be better to do a fresh PR. this will continue to send tons of messages to all already subscribed.

natexcvi commented:
Reopened [here](https://github.com/elastic/kibana/pull/149405). @shahzad31 anything else I should do for the migration?
